drivers/can/can_stm32.c

```diff
@@ -413,16 +441,34 @@ int can_stm32_send(struct device *dev, const struct zcan_frame *msg,
 		return CAN_TX_BUS_OFF;
 	}
 
-	k_mutex_lock(tx_mutex, K_FOREVER);
+	time_stamp = k_uptime_get();
+
+	if (k_mutex_lock(tx_mutex, mb_timeout)) {
+		LOG_DBG("Timeout waiting for tx_mutex");
+		return CAN_TIMEOUT;
+	}
+
 	while (!(transmit_status_register & CAN_TSR_TME)) {
 		k_mutex_unlock(tx_mutex);
 		LOG_DBG("Transmit buffer full. Wait with timeout (%dms)",
-			    timeout);
-		if (k_sem_take(&data->tx_int_sem, timeout)) {
+			    mb_timeout);
+		time_spent = k_uptime_get() - time_stamp;
+		if (time_spent >= mb_timeout || mb_timeout == K_NO_WAIT ||
+		    k_sem_take(&data->tx_int_sem, mb_timeout - time_spent)) {
```

Comment:
I'm not sure if `mb_timeout - time_spent` is allowed as timeout parameter here since I assume milliseconds and not use K_MSEC(). Maybe @pabigot can clarify.

Reply:
This is exactly the sort of code that can't be automatically converted without risk.  The question is whether the driver wants to maintain `mb_timeout` as a count of milliseconds, or as a value to be passed to the kernel API.

Looking at the way the driver seems to do things, and assuming there's no need for timeouts measured in microseconds or anything finer than a millisecond, I would say that the variables used to pass timeouts should remain integers, documented as being specified in milliseconds.

In this code that would mean that `mb_timeout == K_NO_WAIT` would become `mb_timeout == 0`, and the value passed to `k_sem_take` would be `K_MSEC(mb_timeout - time_spent)`.